demos/jobs/jobs_demo_mosquitto/jobs_demo_mosquitto.c

```diff
@@ -0,0 +1,1304 @@
+        default:
+
+            /* process exit status 0 means success */
+            if( ( pid == h->child ) &&
+                ( WIFEXITED( status ) ) &&
+                ( WEXITSTATUS( status ) == 0 ) )
+            {
+                info( "completed job id: %s", h->jobid );
+                report = makeReport_( "SUCCEEDED" );
+            }
+            else
+            {
+                info( "failed job id: %s", h->jobid );
+                report = makeReport_( "FAILED" );
+            }
+
+            h->child = 0;
+            h->runStatus = None;
+    }
+
+    return sendUpdate( h, h->jobid, h->jobidLength, report );
+}
+
+/*-----------------------------------------------------------*/
+
+static bool download( handle_t * h )
+{
+    bool ret = true;
+    pid_t pid;
+
+    assert( h != NULL );
+    assert( h->jobid != NULL );
+    assert( h->url != NULL );
+
+    /* run download as a separate process */
+    pid = fork();
+
+#define dir_format    "%s/job-%s.XXXXXX"
+
+    if( pid == 0 )
+    {
+        /* create a unique download directory */
+        char dir_name[ sizeof( DESTINATION_PREFIX ) + h->jobidLength + sizeof( dir_format ) ];
+
+        snprintf( dir_name, sizeof( dir_name ), dir_format, DESTINATION_PREFIX, h->jobid );
+
+        if( mkdtemp( dir_name ) == NULL )
+        {
+            err( 1, "mkdtemp %s", dir_name );
+        }
+
+        if( chdir( dir_name ) == -1 )
+        {
+            err( 1, "chdir %s", dir_name );
+        }
+
+        info( "download directory: %s", dir_name );
+
+        /* exec the download program */
+        CURL( h->url );
+        /* arrive here only if exec failed */
+        err( 1, "execl:" );
+    }
+
+    free( h->url );
+    h->url = NULL;
+    h->child = pid;
+
+    if( pid == -1 )
+    {
+        warn( "fork" );
+        ret = false;
+    }
+
+    return ret;
+}
+
+/*-----------------------------------------------------------*/
+
+static void cancelDownload( handle_t * h )
+{
+    assert( h != NULL );
+
+    if( h->child > 0 )
+    {
+        int ret = kill( h->child, SIGKILL );
+        assert( ( ret == 0 ) || ( errno == ESRCH ) );
+
+        if( ret == 0 )
+        {
+            /* discard the download process exit status */
+            ( void ) wait( NULL );
+        }
+    }
+
+    h->child = 0;
+}
+
+/*-----------------------------------------------------------*/
+
+static void on_log( struct mosquitto * m,
+                    void * p,
+                    int level,
+                    const char * log )
+{
+    assert( log != NULL );
+
+    info( "%s", log );
+}
+
+/*-----------------------------------------------------------*/
+
+static void catch( int signal )
+{
+    errx( 1, "exit on signal: %d", signal );
+}
+
+/*-----------------------------------------------------------*/
+
+static bool setup( handle_t * h )
+{
+    bool ret = false;
+    struct sigaction sa = { 0 };
+
+    assert( h != NULL );
+
+    /* ensure teardown() will run for these signals */
```

Comment:
teardown() or catch()

Reply:
Specifically, teardown().  Catching the signals means the function registered with on_exit() will run.